src/app/src/main/kotlin/org/vaccineimpact/api/app/repositories/jooq/JooqTouchstoneRepository.kt

```diff
@@ -191,14 +193,62 @@ class JooqTouchstoneRepository(
         return records.map { mapCoverageSet(it) }
     }
 
+    private fun coverageDimensions() : Array<Field<*>>
+    {
+        //The columns in the Coverage table which the target and coverage values are grouped by
+        return arrayOf(COVERAGE.COVERAGE_SET,
+                COVERAGE.YEAR,
+                COVERAGE.COUNTRY,
+                COVERAGE.AGE_FROM,
+                COVERAGE.AGE_TO,
+                COVERAGE.AGE_RANGE_VERBATIM,
+                COVERAGE.GAVI_SUPPORT,
+                COVERAGE.GENDER)
+    }
+
+    private fun targetValid() : Field<BigDecimal?>
+    {
+        //Coverage and target values only make sense if both are provided. Return the target value of a row providing
+        //coverage is not null, else return null
+        return `when`(COVERAGE.COVERAGE_.isNull(), COVERAGE.COVERAGE_) //This is actually the easiest way to get a known NULL here!
+                .otherwise(COVERAGE.TARGET)
+    }
+
+    private fun coverageValid() : Field<BigDecimal?>
+    {
+        //Coverage and target values only make sense if both are provided. Return the coverage value of a row providing
+        //target is not null, else return null
+        return `when`(COVERAGE.TARGET.isNull(), COVERAGE.TARGET) //This is actually the easiest way to get a known NULL here!
+                .otherwise(COVERAGE.COVERAGE_)
+    }
+
+    private fun aggregatedCoverageValues() : List<Field<*>>
+    {
+        return arrayOf(
+                //Aggregated coverage - the sum of each row's coverage * target (to get total no of fvp's across campaign)
+                //divided by the the total target population
+                //Danger of divide by zero error here - treat as NULL if summed target is 0
+                //This is rounded to 2 dec places
+                round(
```

Comment:
Turning zeros into nulls is a problem unfortunately. In the case of routine coverage (so where `COVERAGE_SET.ACTIVITY_TYPE == 'routine'`), all target values will (legitimately) be zero. This FVP calculation won't be appropriate there - we'll have to just return coverage values as is.